File: src/platform/identity/principal.ts
import type { IPrincipal, IClaims } from "@fios/kernel";
import { buildClaims } from "./claims";

export function createPrincipal(input: {
  userId: string;
  username: string;
  role: string;
  tenantId: string;
  companyId?: string | null;
  sessionId?: string;
  permissions?: string[];
}): IPrincipal {
  const claims: IClaims = buildClaims({
    userId: input.userId,
    username: input.username,
    role: input.role,
    tenantId: input.tenantId,
    companyId: input.companyId,
    sessionId: input.sessionId,
    permissions: input.permissions,
  });

  return {
    userId: input.userId,
    username: input.username,
    role: input.role,
    tenantId: input.tenantId,
    companyId: input.companyId ?? null,
    sessionId: input.sessionId,
    claims,
  };
}

export function isPrincipalAuthenticated(principal: IPrincipal | null): boolean {
  return Boolean(principal?.userId && principal?.tenantId);
}

export function principalEquals(a: IPrincipal | null, b: IPrincipal | null): boolean {
  if (!a || !b) return false;
  return a.userId === b.userId && a.tenantId === b.tenantId && a.sessionId === b.sessionId;
}

File: erp_bot/src/oip/infrastructure/security/principal.py
"""Authenticated security principal."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum


class AuthMethod(str, Enum):
    JWT = "jwt"
    API_KEY = "api_key"
    SERVICE_ACCOUNT = "service_account"


@dataclass(frozen=True)
class SecurityPrincipal:
    user_id: str
    tenant_id: str
    company_id: str
    branch_id: str | None
    role: str
    permissions: tuple[str, ...]
    session_id: str
    auth_method: AuthMethod
    username: str = ""
    service_account_id: str | None = None
    roles: tuple[str, ...] = field(default_factory=tuple)

    def has_permission(self, permission: str) -> bool:
        if permission in self.permissions:
            return True
        prefix = permission.split(":")[0]
        return f"{prefix}:*" in self.permissions or "oip:*" in self.permissions or "erp:*" in self.permissions

    def has_any_permission(self, required: tuple[str, ...]) -> bool:
        if not required:
            return True
        return any(self.has_permission(item) for item in required)
